File: mcp_server.py
#!/usr/bin/env python3
"""
Simple Copilot MCP (Model Context Protocol) Server

This is a basic implementation of an MCP server that demonstrates
the core concepts of the Model Context Protocol for GitHub Copilot.
"""

import json
import sys
from typing import Dict, List, Any


class MCPServer:
    """A simple MCP server implementation."""
    
    def __init__(self, name: str = "simple-mcp-server", version: str = "1.0.0"):
        self.name = name
        self.version = version
        self.tools = {}
        self.resources = {}
        
    def register_tool(self, name: str, description: str, handler):
        """Register a tool with the MCP server."""
        self.tools[name] = {
            "name": name,
            "description": description,
            "handler": handler
        }
        
    def register_resource(self, uri: str, name: str, description: str, content: str):
        """Register a resource with the MCP server."""
        self.resources[uri] = {
            "uri": uri,
            "name": name,
            "description": description,
            "content": content
        }
    
    def handle_request(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Handle an incoming MCP request."""
        method = request.get("method")
        params = request.get("params", {})
        
        if method == "initialize":
            return self._handle_initialize(params)
        elif method == "tools/list":
            return self._handle_list_tools()
        elif method == "tools/call":
            return self._handle_call_tool(params)
        elif method == "resources/list":
            return self._handle_list_resources()
        elif method == "resources/read":
            return self._handle_read_resource(params)
        else:
            return {"error": f"Unknown method: {method}"}
    
    def _handle_initialize(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Handle initialization request."""
        return {
            "protocolVersion": "1.0",
            "serverInfo": {
                "name": self.name,
                "version": self.version
            },
            "capabilities": {
                "tools": {},
                "resources": {}
            }
        }
    
    def _handle_list_tools(self) -> Dict[str, Any]:
        """List available tools."""
        return {
            "tools": [
                {
                    "name": tool["name"],
                    "description": tool["description"]
                }
                for tool in self.tools.values()
            ]
        }
    
    def _handle_call_tool(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Call a registered tool."""
        tool_name = params.get("name")
        arguments = params.get("arguments", {})
        
        if tool_name not in self.tools:
            return {"error": f"Tool not found: {tool_name}"}
        
        try:
            result = self.tools[tool_name]["handler"](arguments)
            return {"content": [{"type": "text", "text": result}]}
        except Exception as e:
            return {"error": str(e)}
    
    def _handle_list_resources(self) -> Dict[str, Any]:
        """List available resources."""
        return {
            "resources": [
                {
                    "uri": resource["uri"],
                    "name": resource["name"],
                    "description": resource["description"]
                }
                for resource in self.resources.values()
            ]
        }
    
    def _handle_read_resource(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Read a resource."""
        uri = params.get("uri")
        
        if uri not in self.resources:
            return {"error": f"Resource not found: {uri}"}
        
        resource = self.resources[uri]
        return {
            "contents": [
                {
                    "uri": resource["uri"],
                    "mimeType": "text/plain",
                    "text": resource["content"]
                }
            ]
        }
    
    def run(self):
        """Run the MCP server (stdio mode)."""
        print(f"Starting {self.name} v{self.version}...", file=sys.stderr)
        
        for line in sys.stdin:
            try:
                request = json.loads(line.strip())
                response = self.handle_request(request)
                print(json.dumps(response))
                sys.stdout.flush()
            except json.JSONDecodeError as e:
                error_response = {"error": f"Invalid JSON: {str(e)}"}
                print(json.dumps(error_response))
                sys.stdout.flush()
            except Exception as e:
                error_response = {"error": f"Server error: {str(e)}"}
                print(json.dumps(error_response))
                sys.stdout.flush()


def example_tool_handler(args: Dict[str, Any]) -> str:
    """Example tool handler - echoes back the input."""
    message = args.get("message", "Hello from MCP!")
    return f"Echo: {message}"


def main():
    """Main entry point for the MCP server."""
    # Create and configure the server
    server = MCPServer(name="simple-mcp-server", version="1.0.0")
    
    # Register example tool
    server.register_tool(
        name="echo",
        description="Echoes back the provided message",
        handler=example_tool_handler
    )
    
    # Register example resource
    server.register_resource(
        uri="file:///example.txt",
        name="Example Resource",
        description="An example resource for demonstration",
        content="This is an example resource content."
    )
    
    # Run the server
    server.run()


if __name__ == "__main__":
    main()

File: mcp-server.js
#!/usr/bin/env node

/**
 * Simple Copilot MCP (Model Context Protocol) Server
 * 
 * This server provides basic tools and resources that can be used
 * by AI assistants through the Model Context Protocol.
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
  ListResourcesRequestSchema,
  ReadResourceRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";

/**
 * Create and configure the MCP server
 */
class SimpleMCPServer {
  constructor() {
    this.server = new Server(
      {
        name: "simple-copilot-mcp-server",
        version: "1.0.0",
      },
      {
        capabilities: {
          tools: {},
          resources: {},
        },
      }
    );

    this.setupHandlers();
  }

  /**
   * Setup request handlers for the MCP server
   */
  setupHandlers() {
    // List available tools
    this.server.setRequestHandler(ListToolsRequestSchema, async () => ({
      tools: [
        {
          name: "echo",
          description: "Echoes back the provided message",
          inputSchema: {
            type: "object",
            properties: {
              message: {
                type: "string",
                description: "The message to echo back",
              },
            },
            required: ["message"],
          },
        },
        {
          name: "calculate",
          description: "Performs basic arithmetic operations (add, subtract, multiply, divide)",
          inputSchema: {
            type: "object",
            properties: {
              operation: {
                type: "string",
                enum: ["add", "subtract", "multiply", "divide"],
                description: "The arithmetic operation to perform",
              },
              a: {
                type: "number",
                description: "First number",
              },
              b: {
                type: "number",
                description: "Second number",
              },
            },
            required: ["operation", "a", "b"],
          },
        },
        {
          name: "get_timestamp",
          description: "Returns the current timestamp",
          inputSchema: {
            type: "object",
            properties: {
              format: {
                type: "string",
                enum: ["iso", "unix", "locale"],
                description: "Format of the timestamp (default: iso)",
              },
            },
          },
        },
      ],
    }));

    // Handle tool calls
    this.server.setRequestHandler(CallToolRequestSchema, async (request) => {
      const { name, arguments: args } = request.params;

      try {
        switch (name) {
          case "echo":
            return {
              content: [
                {
                  type: "text",
                  text: `Echo: ${args.message}`,
                },
              ],
            };

          case "calculate":
            const { operation, a, b } = args;
            let result;
            
            switch (operation) {
              case "add":
                result = a + b;
                break;
              case "subtract":
                result = a - b;
                break;
              case "multiply":
                result = a * b;
                break;
              case "divide":
                if (b === 0) {
                  throw new Error("Division by zero is not allowed");
                }
                result = a / b;
                break;
              default:
                throw new Error(`Unknown operation: ${operation}`);
            }

            return {
              content: [
                {
                  type: "text",
                  text: `Result: ${a} ${operation} ${b} = ${result}`,
                },
              ],
            };

          case "get_timestamp":
            const format = args.format || "iso";
            const now = new Date();
            let timestamp;

            switch (format) {
              case "iso":
                timestamp = now.toISOString();
                break;
              case "unix":
                timestamp = Math.floor(now.getTime() / 1000).toString();
                break;
              case "locale":
                timestamp = now.toLocaleString();
                break;
              default:
                timestamp = now.toISOString();
            }

            return {
              content: [
                {
                  type: "text",
                  text: `Current timestamp (${format}): ${timestamp}`,
                },
              ],
            };

          default:
            throw new Error(`Unknown tool: ${name}`);
        }
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error: ${error.message}`,
            },
          ],
          isError: true,
        };
      }
    });

    // List available resources
    this.server.setRequestHandler(ListResourcesRequestSchema, async () => ({
      resources: [
        {
          uri: "info://server",
          name: "Server Information",
          description: "Information about this MCP server",
          mimeType: "text/plain",
        },
        {
          uri: "info://capabilities",
          name: "Server Capabilities",
          description: "List of server capabilities",
          mimeType: "application/json",
        },
      ],
    }));

    // Handle resource reads
    this.server.setRequestHandler(ReadResourceRequestSchema, async (request) => {
      const { uri } = request.params;

      switch (uri) {
        case "info://server":
          return {
            contents: [
              {
                uri,
                mimeType: "text/plain",
                text: "Simple Copilot MCP Server v1.0.0\n\n" +
                      "This is a basic MCP server implementation that provides:\n" +
                      "- Echo tool: Echoes back messages\n" +
                      "- Calculate tool: Performs basic arithmetic\n" +
                      "- Get timestamp tool: Returns current time in various formats\n\n" +
                      "For more information, visit the repository.",
              },
            ],
          };

        case "info://capabilities":
          return {
            contents: [
              {
                uri,
                mimeType: "application/json",
                text: JSON.stringify(
                  {
                    tools: ["echo", "calculate", "get_timestamp"],
                    resources: ["info://server", "info://capabilities"],
                    version: "1.0.0",
                  },
                  null,
                  2
                ),
              },
            ],
          };

        default:
          throw new Error(`Unknown resource: ${uri}`);
      }
    });
  }

  /**
   * Start the MCP server
   */
  async run() {
    const transport = new StdioServerTransport();
    await this.server.connect(transport);
    console.error("Simple Copilot MCP Server running on stdio");
  }
}

// Start the server
const server = new SimpleMCPServer();
server.run().catch((error) => {
  console.error("Server error:", error);
  process.exit(1);
});
